hibernate-core/src/main/java/org/hibernate/internal/SessionFactoryImpl.java

```diff
@@ -260,7 +265,7 @@ public SessionFactoryImpl(
 		LOG.debugf( "Session factory constructed with filter configurations : %s", filters );
 		LOG.debugf( "Instantiating session factory with properties: %s", properties );
 
-		this.queryPlanCache = new QueryPlanCache( this );
+		this.queryPlanCache = queryPlanCacheFunction.apply( this );
```

Comment:
@DavideD I'm not a big fan of this. I think it might be better to add a `getQueryPlanCache()` accessor to `SessionFactoryImpl` and override it on the subclass.

Reply:
(You're passing a reference to a partially-uninitialized object to an arbitrary function. The subclass constructor hasn't been called at this point.)

Reply:
And note: you're using a function ref here precisely because you want to circumvent the JVM's restriction which is designed to prevent you doing precisely what you're doing here :-)

Reply:
You got me :) OK, I will change it